### PR TITLE
add second lifetime to `FromPyObject`

### PR DESCRIPTION
**Pull request by Icxolu**
This adds the second lifetime to `FromPyObject` and removes `FromPyObjectBound`. I hope I adjusted all the bounds correctly. I believe the equivalent to our current implementation is use HRTB for the input lifetime on any generic bound. But this should only be necessary for containers that create temporary Python references during extraction. For wrapper types it should be possible to just forward the relaxed bound.

For easier trait bounds `FromPyObjectOwned` is introduced. This is blanket implemented for any `FromPyObject` type that does not depend on the input lifetime. It is intended to be used as a trait bound, the idea is inspired by `serde` (`Deserialize` <=> `DeserializeOwned`)

I tried to document different cases in the migration guide, but it can probably still be extended. ~Changelog entry is still missing, will write that tomorrow.~

Breaking changes:
- the second lifetime
- the `extract` method without default
- the addition lifetime bound on `extract_bound` to make the default work

### CODE REVIEW

**Comment by mejrs** [comment]
I propose that we add a `FromPyObjectOwned` trait similar to serde's [DeserializeOwned](https://docs.rs/serde/latest/serde/de/trait.DeserializeOwned.html). It's much easier to explain "just use this trait if the output doesn't borrow from Python" rather than "you need higher ranked trait bounds".

**Comment by Icxolu** [comment]
Thanks! I think that's a great idea. It's essentially the same thing, but it's much easier to grasp by giving it a distinct name, plus we can better document it in the API docs. I will adapt this to make use of that.

File: src/conversion.rs [hunk 1]
```diff
@@ -336,30 +296,30 @@ where
...
     T: PyClass,
 {
-    fn extract_bound(obj: &Bound<'py, PyAny>) -> PyResult<Self> {
+    fn extract(obj: Borrowed<'_, 'py, PyAny>) -> PyResult<Self> {
         obj.downcast::<T>()?.try_borrow().map_err(Into::into)
     }
 }
 
-impl<'py, T> FromPyObject<'py> for PyRefMut<'py, T>
+impl<'py, T> FromPyObject<'_, 'py> for PyRefMut<'py, T>
```
Comment (davidhewitt) [inline src/conversion.rs:318]: And then the follow-up future PR will be what to do about `PyRef`. Should we split into `PyRef` / `PyRefBorrowed`, or keep the one type `PyRef<'a, 'py, T>` and accept some breakage? 🤔

**Review by davidhewitt — COMMENTED** [review]
Thanks for pushing on with this! Overall it looks good to me, and I only have a few docs suggestions.

Before we definitely commit to this, I want to take a brief moment just to reflect on the choice of `Borrowed` over `&Bound`. There's at least two good technical reasons, which is the lifetime constraint and performance by avoiding pointer-to-pointer. The codspeed branch does show a few slight perf improvements in the ~4% range on tuple extraction and we have further possibilities like `PyRef` containing `Borrowed` once we do this.

That said, we deliberately kept the `Borrowed` smart pointer out of the way as much as possible in the original 0.21 design to try to keep new concepts down. Is it fine to increase its use? I _think_ the upsides justify this, though we might want to increase the quality of documentation and examples on `Borrowed` (which I'd left quite minimal so far).

File: pyo3-macros-backend/src/frompyobject.rs [hunk 2]
```diff
@@ -616,8 +616,9 @@ pub fn build_derive_from_pyobject(tokens: &DeriveInput) -> Result<TokenStream> {
     let ident = &tokens.ident;
     Ok(quote!(
         #[automatically_derived]
-        impl #trait_generics #pyo3_path::FromPyObject<#lt_param> for #ident #generics #where_clause {
-            fn extract_bound(obj: &#pyo3_path::Bound<#lt_param, #pyo3_path::PyAny>) -> #pyo3_path::PyResult<Self>  {
+        impl #trait_generics #pyo3_path::FromPyObject<'_, #lt_param> for #ident #generics #where_clause {
```
[the remark this answers is not in the document]
Comment (Icxolu) [reply pyo3-macros-backend/src/frompyobject.rs:619]: Yeah, something to experiment with in the future, but out of scope for now for sure 😅

File: src/conversion.rs [hunk 3]
```diff
@@ -198,90 +197,27 @@ pub trait IntoPy<T>: Sized {
...
-/// be expressed by the existing `FromPyObject` trait once the GIL Refs API was
-/// removed.
-///
-/// # Usage
-///
-/// Users are prevented from implementing this trait, instead they should implement
-/// the normal `FromPyObject` trait. This trait has a blanket implementation
-/// for `T: FromPyObject`.
+/// Note: depending on the implementation, the lifetime of the extracted result may
+/// depend on the lifetime of the `obj` or the `prepared` variable.
```
[the remark this answers is not in the document]
Comment (Icxolu) [reply src/conversion.rs:201]: I agree, I was a bit lazy and just uncommented the docs that we temporarily removed during gil-ref migration.

How about
```rust
/// Note: depending on the implementation, the extracted result may
/// depend on the Python lifetime `'py` or the input lifetime `'a` of `obj`.
```

Also the example below that seem a bit confusing, what do you think about using `Cow<str>` as an example, which may or may not borrow from the input (`'a`) depending on the Python runtime type and maybe `Bound<'py, PyString>` as an example that depends on the Python lifetime? Maybe we also add an example of a collection type (maybe `Vec`) to introduce `FromPyObjectOwned`...

File: src/conversion.rs (hunk 1 above)
Comment (Icxolu) [reply src/conversion.rs:318]: Hmm, I think I we want to provide both variants, I would prefer `PyRef<'a, 'py, T>` and `PyRefOwned<'py, T>` since that would be more consistent with `FromPyObjectOwned`. 

In any case we would need to find a good way to differentiate their constructors. Neither `obj.borrow_borrowed()` nor `obj.borrow_owned()` feel to appealing to me 😄

**Comment by Icxolu** [comment]
Thanks for the review! You are right, the choice of `Borrowed` over `&Bound` is definitely something to discuss about. I used it here because it was the final form of `FromPyObjectBound` that we landed on. Looking back, the change to `Borrowed` happened in #3959, which was related to the gil-refs API, so it might not be strictly necessary anymore (haven't checked). The points you mentioned seem like good justification why it still makes sense to introduce additional complexity here.

> I _think_ the upsides justify this, though we might want to increase the quality of documentation and examples on `Borrowed` (which I'd left quite minimal so far).

I agree, if we put `Borrowed` in such a core part of the infrastructure it deserves a documentation overhaul. We might also want to take a look at the current (public) API and additionally expose some helpers. Something like `to_any` comes to mind, which is currently internal only.

File: src/conversion.rs (hunk 1 above)
Comment (davidhewitt) [reply src/conversion.rs:382]: Agreed, it's tough to find names. I'm somewhat tempted to wait until after 0.23 and to first see how all these trait changes we've already made play out before we change `PyRef`.

File: src/conversion.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (davidhewitt) [reply src/conversion.rs:282]: I think all of that sounds great, yes please 👍

**Review by mejrs — CHANGES_REQUESTED** [review]
IMO the recently added `FromPyObject` documentation is too "implementation detail"-y.

I'd prefer if this documentation is kept short and concise, one paragraph max.

For example:

```
Depending on the python version and implementation some `FromPyObject`
 implementations may produce Rust types that point into the Python type.
For example `PyString` and `PyBytes` may convert into `str` and `[u8]` references
 that borrow from the original Python object.

Types that do not borrow from the input should use `FromPyObjectOwned` instead.
```

**Comment by Icxolu** [comment]
Thanks for the feedback! I took another pass over it and tried to make it a bit more concise. I do however think that it's worth to talk about the individual lifetimes at play here, since this trait is right in the center of PyO3. I moved the paragraph about collections into the `FromPyObjectOwned` docs, because it works much better there with the code example, and replaced it with a smaller note for more info. I reworded my examples a bit and moved them under a `# Details` section, maybe this works better?

Let me know what you think about this.

File: src/conversion.rs [hunk 4]
```diff
@@ -278,21 +279,44 @@ impl<'a, 'py, T> IntoPyObject<'py> for &'a Py<T> {
...
-/// Since which case applies depends on the runtime type of the Python object,
-/// both the `obj` and `prepared` variables must outlive the resulting string slice.
+/// Note: depending on the implementation, the extracted result may depend on
+/// the Python lifetime `'py` or the input lifetime `'a` of `obj`.
+///
+/// For example, when extracting a [`Cow<'a, str>`] the result may or may not
+/// borrow from the input lifetime `'a`. The behavior depends on the runtime
+/// type of the Python object. For a Python byte string, the existing string
+/// data can be borrowed (lifetime: `'a`) into a [`Cow::Borrowed`]. For a Python
+/// Unicode string, the data may have to be reencoded to UTF-8, and copied into
```
[the remark this answers is not in the document]
Comment (Icxolu) [reply src/conversion.rs]: Yeah, I'm aware of that, but that felt like too much detail for me, and is not really relevant for the point I am trying to make here. That's why I tried to phrase it in a rather vague way.

**Review by davidhewitt — APPROVED** [review]
Thanks very much for keeping this branch alive, I guess let's move forward with this before it gets too painful to keep updating. 

I had yet another reflection on `&Bound` vs `Borrowed` and I'm feeling reasonably comfortable that `Borrowed` is the correct choice.

**Comment by Icxolu** [comment]
I _believe_ this now also finishes the `_bound` methods deprecations 🎉

**Comment by davidhewitt** [comment]
Fantastic, thank you so much for your help on pushing that over the line. When I get some time I'll try to write up a list of what we might have left for 0.23 (hopefully not much).

**Comment by davidhewitt** [comment]
Ah, I just started playing with this locally and found one additional finding: having `Borrowed<'a, 'py, T>` as the input type means that it's impossible to implement `FromPyObject<'a, 'py> for &'a Bound<'py, T>` - because the input borrowed can't be cast into a `&'a Bound<'py, T>` for a sufficient lifetime (pointer goes in, pointer-to-pointer goes out).

Is that a problem? I think not, because that's what `.downcast()` is for (and it's potentially nice to separate them like this), however just worth a brief double-check that others agree.

**Comment by davidhewitt** [comment]
Oh, and one more case which comes up in local testing. At the moment `from_py_with` extractors take `&Bound<'py, PyAny>`. Should we be considering changing those to `Borrowed` too, for consistency?

**Comment by davidhewitt** [comment]
> Should we be considering changing those to Borrowed too, for consistency?

It feels to me like we might want to consider macro tricks such that users can use both `&Bound` and `Borrowed`.

**Comment by Icxolu** [comment]
> having `Borrowed<'a, 'py, T>` as the input type means that it's impossible to implement `FromPyObject<'a, 'py> for &'a Bound<'py, T>`

That's true, but as you mention below, I actually think this a good thing and makes a clearer use case between `.extract()` and `.downcast()`.

> Oh, and one more case which comes up in local testing. At the moment `from_py_with` extractors take `&Bound<'py, PyAny>`. Should we be considering changing those to `Borrowed` too, for consistency?

Hmm, good question. I guess that is part of the question about how prominent we want `Borrowed` to be. Providing some flexibility via macro magic as you suggested sounds like a good compromise. I can explore that as a followup.

**Comment by davidhewitt** [comment]
Yes agreed, it's definitely a follow-up to make changes to expand `from_py_with`.

I wonder, are there any other options we have here? E.g. does it work to have a trait with both `extract` and `extract,_borroeed`, maybe with a default implementation of `extract,_borrowed`? I suspect that might not work because of the same reason we can't currently have `&Bound` as an output.

I am still feeling a little uneasy about the choice to commit `Borrowed` more directly into the face of users. I think it is probably the right choice still, as long as we improve the documentation and API on borrowed. Maybe I will quickly try to get the perf wins out of `extract_argument.rs` later today to add another data point.

**Comment by davidhewitt** [comment]
And (sorry to keep dangling ideas on here), inspired by the new `IntoPyObject`, I have two further perf related thoughts while we are making breaking changes here:

- In the past we have noted that the `PyResult` return type forced conversion of errors to `PyErr` and made `.extract()` a performance footgun compared to `.downcast()`. There is still this problem for `FromPyObject for Bound<T>`. Should we add a `type Error` in the same way we are adding for `IntoPyObject`?
- It is possible to call `.extract()` on e.g. `Bound<PyString>` to get a `String`, but this comes at the cost of a redundant Python type check because `FromPyObject` always works on `PyAny`. Could we add a `type Source`, and add a way to make `Bound<T>` directly offer `extract()` where `U: FromPyObject<Source = T>`?

Both changes would significantly complicate the trait. I wonder, is there a way that we can support a simple trait like `FromPyObject` and a more advanced `FromPyObjectImpl`, which has a blanket from `FromPyObject`, so users can just implement the simple trait if that's good enough?

All food for thought and I think worth discussing while we're committing to breaking changes here.

**Comment by Icxolu** [comment]
> And (sorry to keep dangling ideas on here), inspired by the new `IntoPyObject`, I have two further perf related thoughts while we are making breaking changes here:

No problem 😄 

> * In the past we have noted that the `PyResult` return type forced conversion of errors to `PyErr` and made `.extract()` a performance footgun compared to `.downcast()`. There is still this problem for `FromPyObject for Bound<T>`. Should we add a `type Error` in the same way we are adding for `IntoPyObject`?

If we're willing to do more breaking changes here, that sounds like a good idea to me. The only downsides I can see
- sadly we can't provide a default type
- similar to `IntoPyObject` we would want `FromPyObject::Error: Into<PyErr>`, so this has the same effect on generic code and may tie into the same proposal you raised in https://github.com/PyO3/pyo3/pull/4489#discussion_r1730272529

> * It is possible to call `.extract()` on e.g. `Bound<PyString>` to get a `String`, but this comes at the cost of a redundant Python type check because `FromPyObject` always works on `PyAny`. Could we add a `type Source`, and add a way to make `Bound<T>` directly offer `extract()` where `U: FromPyObject<Source = T>`?

This would be really cool indeed. I guess we would maybe somehow need to take advantage of method precedence between trait and inherent methods, to make the fall through to `PyAny` work...

**Comment by davidhewitt** [comment]
If we're breaking `FromPyObject`, I wonder if we can solve #2888 similar to how we handled bytes in `IntoPyObject` 🤔

**Comment by Icxolu** [comment]
That would make a very nice symmetry. I think if we do it, then it should work for all containers, which would probably mean one hidden method per container (vec, smallvec, arrays), maybe these could be unified with "return position impl trait in trait" (with something like `impl Iterator`) once MSRV allows. The more challenging thing seems to be to restrict a default implementation in `FromPyObject` to `FromPyObjectOwned`...

**Comment by Icxolu** [comment]
> If we're breaking `FromPyObject`, I wonder if we can solve #2888 similar to how we handled bytes in `IntoPyObject` 🤔

I've looked into this again. I think we would (for example) want something like this
```rust
trait FromPyObject<'a, 'py>: Sized {
   // snip

    #[doc(hidden)]
    fn extract_array<const N: usize>(
        obj: Borrowed<'_, 'py, PyAny>,
        _: private::Token,
    ) -> PyResult<[Self; N]>
    where
        Self: FromPyObjectOwned<'py>,
    {
        todo!()
    }
}
```
which we then use inside of `extract` on the array implementation and overwrite on the `u8` implementation.

Unfortunately this currently does not compile. I believe this is the corresponding `rustc` issue: https://github.com/rust-lang/rust/issues/34979, which refers to the next trait solver as the solution and `-Znext-solver` does indeed make this work. Unless there is something I have overseen, I think we have to wait with the bytes specialization until `-Znext-solver` stabilizes.

**Comment by davidhewitt** [comment]
Ah, nice research 👍. Agreed we can make a reminder for ourselves to improve bytes specializations once MSRV uses the next solve (gonna be a while 😂).

Given the early reception of `IntoPyObject` seems to be that it's hard, I'm currently feeling like we would build up some goodwill by being gentler on users for a few releases so am feeling like backing down on my other proposals for advanced optimisations above. We can introduce them incrementally in the future perhaps.

This branch really deserves to be merged asap, so please forgive me the delay, I think this now is highest thing on my to-do perhaps excluding a 0.23.1 patch release. I will try to play around with this branch and see how I feel about the ergonomics & perf, maybe once the kids are in bed tonight.

**Comment by Icxolu** [comment]
> Given the early reception of `IntoPyObject` seems to be that it's hard, I'm currently feeling like we would build up some goodwill by being gentler on users for a few releases so am feeling like backing down on my other proposals for advanced optimisations above. We can introduce them incrementally in the future perhaps.

I tend to agree here. We can maybe think about adding the `Error` type. That seem pretty low hanging and fairly intuitive. 

> This branch really deserves to be merged asap, so please forgive me the delay,

Absolutely no worries :)

**Comment by davidhewitt** [comment]
One idea which struck me, what if we keep `FromPyObject` with just a single lifetime and `&Bound` as input, and we instead expose what's currently called `FromPyObjectBound` as `FromPyObjectBorrowed`?

The idea would be that here we are currently breaking the main trait have two lifetimes, and we introduce `FromPyObjectOwned` for simpler trait bounds, which kinda makes sense because that matches `serde`'s `DeserializeOwned`.

But I think unlike `serde`'s `DeserializeOwned`, most uses of `FromPyObject` do not want care about the second lifetime, so maybe it makes sense to keep that to a specialised trait .

This might have the nice boost that it's probably less breaking than adding the second lifetime to `FromPyObject`.

I feel like there's some detail about `PyRef` and second lifetimes which I might have overlooked with this idea. But it still seems interesting to think about further.

**Comment by Icxolu** [comment]
Interesting thought, I have't played around with it yet, but I opened with #4720 based on this including the changes adding the second lifetime to `PyRef` and `PyRefMut` (was quite a bit of lifetime reviewing until the borrow checker was happy 😅 )

Maybe we can use that to play around with and judge whats the best option here. My initial feeling says that the one single trait approach might have a bigger hurdle in the beginning but is easier in the long run (less traits, using common patters) but I haven't tried the other approach yet.

**Comment by Icxolu** [comment]
I did play around with the idea a bit. It looks like it is possible to stick the `PyRef(Mut)` changes from #4720 onto it without too many changes. However from a usability perspective it doesn't feel that great IMO. With the two trait approach one has to remember to implement `FromPyObject` if possible, because it is more general due to the blanket, but for uses in trait bounds `FromPyObjectBorrowed` should be used (if possible), because it includes more types. So by default one would need two different traits, depending on the context it is used in. I think there is a good chance that users will get this wrong and use a more restrictive bound than necessary (both in `impl`s and `trait` bounds)

**Comment by davidhewitt** [comment]
Hmm, interesting. One nice thing about

> one has to remember to implement `FromPyObject` if possible

is that I'd guess most users would choose to implement it anyway as it's the simpler to implement. It's quite appealing to me that the default is simple. Plus it's less breaking than adding the lifetime & using `Borrowed` as the input.

I also feel like it's probably quite rare to have a case where `FromPyObjectBorrowed` trait bound is actually critical. At worst case it probably forces an extra clone or reference counting operation if the bound is `FromPyObject` instead? And it's so hard to soundly get borrowed references from python operations (reading from a tuple is possibly the only way?) that probably most cases where a `FromPyObjectBorrowed` bound might actually help, it's probably where the caller already has the reference and could probably just use `.extract()` up front.

I'm quite happy to be wrong here, though my gut feeling is that if cheap-and-cheerful `FromPyObject` works 99% of the time and `FromPyObjectBorrowed` is the power tool to use for edge cases which borrow their input python objects then that might be an easier library for users to learn.

**Comment by Icxolu** [comment]
I though about this some more recently while trying to add `FromPyObject::Error` (independently from this). It turns out that having the traits split (like in `main`) kind of reduces the usefulness of such an error type. Often times the error could be `DowncastError`, but for that we need to have the second lifetime. So we either have to convert to `PyErr`, which we want to avoid, or switch to `FromPyObjectBound` which causes quite a few problem with the trait bounds.

Together with #4965 I think that two separate traits for this is too messy and confusing. The full switch proposed here has less edge cases/is more consistent and is easier to reason about IMO.

**Comment by davidhewitt** [comment]
Sorry to have taken so long to circle back here.

> I though about this some more recently while trying to add `FromPyObject::Error` (independently from this). It turns out that having the traits split (like in main) kind of reduces the usefulness of such an error type. Often times the error could be `DowncastError`, but for that we need to have the second lifetime.

I guess that a new possibility since March might be that we could have use GATs and have `type Error<'a>;`, which should allow the error type to have a second lifetime even if the main trait didn't. But maybe there is a painful mismatch when then trying to pair that with a second trait which has a second lifetime.

> 

> > If we're breaking `FromPyObject`, I wonder if we can solve #2888 similar to how we handled bytes in `IntoPyObject` 🤔
> 
> I've looked into this again. I think we would (for example) want something like this
> 
> ```rust
> trait FromPyObject<'a, 'py>: Sized {
>    // snip
> 
>     #[doc(hidden)]
>     fn extract_array<const N: usize>(
>         obj: Borrowed<'_, 'py, PyAny>,
>         _: private::Token,
>     ) -> PyResult<[Self; N]>
>     where
>         Self: FromPyObjectOwned<'py>,
>     {
>         todo!()
>     }
> }
> ```
> 
> which we then use inside of `extract` on the array implementation and overwrite on the `u8` implementation.
> 
> Unfortunately this currently does not compile. I believe this is the corresponding `rustc` issue: [rust-lang/rust#34979](https://github.com/rust-lang/rust/issues/34979), which refers to the next trait solver as the solution and `-Znext-solver` does indeed make this work. Unless there is something I have overseen, I think we have to wait with the bytes specialization until `-Znext-solver` stabilizes.

I also just had the insight that if we don't have the second lifetime, then it compiles. i.e. this is fine:

```rust
trait FromPyObject<'py>: Sized {
   // snip

    #[doc(hidden)]
    fn extract_array<const N: usize>(
        obj: Borrowed<'_, 'py, PyAny>,
        _: private::Token,
    ) -> PyResult<[Self; N]>
    where
        Self: FromPyObject<'py>,
    {
        todo!()
    }
}
```

... interestingly, it seems like this is also true if we ended up replacing the lifetime as `FromPyObject<'a>` as per your idea in https://github.com/PyO3/pyo3/pull/4720#issuecomment-2495978288

I do quite like `FromPyObject` not having multiple lifetimes from a user simplicity perspective. So I am very tempted to see how this might look now; I have some time to experiment tonight / tomorrow morning.

**Comment by Icxolu** [comment]
> ... interestingly, it seems like this is also true if we ended up replacing the lifetime as `FromPyObject<'a>` as per your idea in [#4720 (comment)](https://github.com/PyO3/pyo3/pull/4720#issuecomment-2495978288)

Another angle on this idea:
We could switch the lifetime like so
```rust
pub trait FromPyObject<'a>: Sized {
    type Error: Into<PyErr>;
    fn extract<'py>(obj: &'a Bound<'py, PyAny>) -> Result<Self, Self::Error>;
}
```
which should be able to support any type (including all which currently need `FromPyObjectBound`) which does not itself depend on `'py`. This leaves mainly `Bound` and `Borrowed` (and `PyRef(Mut)`, but if we replace them with `PyClassGuard(Mut)` as in #5233 this goes away) which we could move to `PyFunctionArgument` like is already done for `&Bound`. The main downsides are
- we would loose the ability to extract them inside of containers i.e. `Vec<Bound<'py, T>>`, given that `Vec<Py<T>>` (or even `Vec<PyClassGuard<'a, T>>` in the future) would still be possible, maybe that's not too bad?
- user defined types could no longer include `Bound`s only `Py`s, but they would gain the ability for borrow from `'a`

Just some thoughts, I'm still not really sure what direction we should take here. I still kind of want to get rid of at least one trait here. Having `PyFunctionArgument`, `FromPyObject` and `FromPyObjectBound` (or equivalent) that all kind of do the same thing makes it kind of hard to follow which types work where, what to implement and what to use as a trait bound...

> I do quite like `FromPyObject` not having multiple lifetimes from a user simplicity perspective.

I generally agree, but on the other hand the double lifetime approach is still the most flexible option, with the least "surprises". Having the functionality split only really hides the complexity for the easy case while often making it worse for the cases that need the additional flexibility. I'm kind of torn here...

**Comment by davidhewitt** [comment]
TLDR; I have thought hard about this over the last week and think this proposed design is the right one and we should move forward with it 👍 

I think given the scale of the break, let's release PyO3 0.26 first ASAP to get the `with_gil` rename shipped and then get this merged. In the meanwhile we can get this PR up-to-date with `main` and give it a few review passes.

----

I've been playing around with this branch and variants like switching the lifetime proposed just above.

The downside of switching the lifetime is it's pretty breaking:
- without reintroducing another trait (`FromPyObjectAttached`) then _all_ extraction of `Bound<'py, T>` straight up breaks
- we still have trouble with containers which then need to implement `FromPyObjectAttached` and then it turns out that `FromPyObject` becomes generally useless as a trait bound because of the same two-trait split issue you highlighted in https://github.com/PyO3/pyo3/pull/4390#issuecomment-2495455971

In general I've come to the same conclusion; a two-trait system is a usability issue, arguably a bit worse even than having the two lifetimes as proposed here.

I'm also reasonably convinced that `Borrowed<'a, 'py, T>` is correct over `&'a Bound<'py, T>`, because otherwise `.extract()` of `&[u8]` from `Py<PyBytes>` (for example) is arbitrarily restricted by implicit `'py: 'a` of the ref-to-bound. Also, we need to use `Borrowed` because otherwise we lose the ability to support tuple extraction.

I also found that this proposal _is_ compatible with optimized `Vec<u8>` extraction - see #5244 

----

So in conclusion:
- I think the trait as implemented here is the right form; it's the only one with full flexibility
- The UX hurdle is the two lifetimes, hopefully we can solve this with good documentation both on the trait and in the migration guide.
- The (smaller) UX hurdle is the use of `Borrowed`; again I think we should just embrace this and put good documentation on the type.

So let's move forward here with this PR? I'm happy to take on a separate PR to add documentation to `Borrowed` 👍

**Comment by davidhewitt** [comment]
(And I'll hope that maybe in the long run the `'py` lifetime might be replaced by more advanced Rust language features so everything simplifies 😂)

**Comment by davidhewitt** [comment]
Ah, and while we're updating this branch I think we should get the `Error` type in.

**Comment by Icxolu** [comment]
I do have a rebased version of this locally, and an additional commit for the error type on top, which I can push here shortly.

File: guide/src/migration.md [hunk 5]
```diff
@@ -3,6 +3,128 @@
 This guide can help you upgrade code through breaking changes from one PyO3 version to the next.
 For a detailed list of all changes, see the [CHANGELOG](changelog.md).
 
+## from 0.26.* to 0.27
+### `FromPyObject` gains additional lifetime
```
Comment (davidhewitt) [inline guide/src/migration.md]: Given there's also the error associated type, it's more than just the lifetime now. Suggestion for the title below, I think we also might want to adjust the text. For the errors I would expect that implementors should be recommended to just use `PyErr` initially and then consider changing that to a more specific error if they want to avoid creating Python exceptions for performance.

```suggestion
### `FromPyObject` reworked for flexibility and efficiency
```

File: guide/src/migration.md [hunk 6]
```diff
@@ -3,6 +3,128 @@
...
+<details open>
+<summary><small>Click to expand</small></summary>
+
+With the removal of the `gil-ref` API it is now possible to fully split the Python GIL lifetime
```
Comment (davidhewitt) [inline guide/src/migration.md]: not sure if I like my suggestion "Python interpreter attach lifetime", we should probably decide on a nice phrase for this 🤔 

```suggestion
With the removal of the `gil-ref` API in PyO3 0.23 it is now possible to fully split the Python interpreter attach lifetime
```

File: guide/src/migration.md [hunk 7]
```diff
@@ -3,6 +3,128 @@
...
+        let mut v = MyVec(Vec::new());
+        for item in obj.try_iter()? {
+            v.0.push(item?.extract::<T>().map_err(Into::into)?);
+        }
+        Ok(v)
+    }
+}
+```
+
+This is very similar to `serde`s `Deserialize` and `DeserializeOwned` traits.
```
Comment (davidhewitt) [inline guide/src/migration.md]: Possibly link to serde docs here?

File: pyo3-benches/benches/bench_dict.rs [hunk 8]
```diff
@@ -62,8 +62,9 @@ fn extract_hashmap(b: &mut Bencher<'_>) {
         let dict = (0..LEN as u64)
             .map(|i| (i, i * 2))
             .into_py_dict(py)
-            .unwrap();
-        b.iter(|| HashMap::<u64, u64>::extract_bound(&dict));
+            .unwrap()
+            .into_any();
```
Comment (davidhewitt) [inline pyo3-benches/benches/bench_dict.rs:66]: I suppose the `.into_any()` was needed here becase `Borrowed<'_, '_, PyDict>` does not automatically coerce to `Borrowed<'_, '_, PyAny>`?

Seems unfortunate but probably fine as most usage should use `.extract()` as a method call where the deref will apply.

File: src/conversion.rs [hunk 9]
```diff
@@ -271,24 +272,30 @@ impl<'py, T> IntoPyObjectExt<'py> for T where T: IntoPyObject<'py> {}
...
+/// `'a` of `obj`.
 ///
-/// During the migration of PyO3 from the "GIL Refs" API to the `Bound<T>` smart pointer, this trait
-/// has two methods `extract` and `extract_bound` which are defaulted to call each other. To avoid
-/// infinite recursion, implementors must implement at least one of these methods. The recommendation
-/// is to implement `extract_bound` and leave `extract` as the default implementation.
-pub trait FromPyObject<'py>: Sized {
+/// Types that must not borrow from the input can use [`FromPyObjectOwned`] as a restriction. This
+/// is most often the case for collection types. See its documentation for more details.
+///
```
Comment (davidhewitt) [inline src/conversion.rs:281]: I wonder if we should have a section to this doc "Implementing `FromPyObject`" which lists the choices available:
 - `#[derive(FromPyObject)]` 
 - manual implementation for types without a lifetime (i.e. `impl FromPyObject<'_, '_> for X`)
 - manual implementation for types with a lifetime

File: src/conversion.rs [hunk 10]
```diff
@@ -271,24 +272,30 @@ impl<'py, T> IntoPyObjectExt<'py> for T where T: IntoPyObject<'py> {}
...
+/// copied into a [`Cow::Owned`]. It does _not_ depend on the Python lifetime `'py`.
+///
+/// The output type may also depend on the Python lifetime `'py`. This allows the output type to
+/// keep interacting with the Python interpreter. See also [`Bound<'py, T>`].
+///
+/// [`Cow<'a, str>`]: std::borrow::Cow
+/// [`Cow::Borrowed`]: std::borrow::Cow::Borrowed
+/// [`Cow::Owned`]: std::borrow::Cow::Owned
+pub trait FromPyObject<'a, 'py>: Sized {
+    /// The type returned in the event of a conversion error.
```
Comment (davidhewitt) [inline src/conversion.rs:371]: I think we should note that for most uses `PyErr` is totally fine, users can provide more specific errors if they potentially want to avoid the cost of creating a Python exception.

File: src/conversions/indexmap.rs [hunk 11]
```diff
@@ -130,17 +130,22 @@ where
...
+
+    fn extract(ob: Borrowed<'_, 'py, PyAny>) -> Result<Self, Self::Error> {
         let dict = ob.cast::<PyDict>()?;
         let mut ret = indexmap::IndexMap::with_capacity_and_hasher(dict.len(), S::default());
-        for (k, v) in dict {
-            ret.insert(k.extract()?, v.extract()?);
+        for (k, v) in dict.iter() {
+            ret.insert(
+                k.extract().map_err(Into::into)?,
+                v.extract().map_err(Into::into)?,
```
Comment (davidhewitt) [inline src/conversions/indexmap.rs:147]: It seems unexpected that the `.map_err` calls were needed here?

```suggestion
                k.extract()?,
                v.extract()?,
```

Oh wait, is this the same sort of problem as we had with `IntoPyObject`, where the `?` desugaring relies on `From`? Sigh. Maybe we just note this somewhere in the docs. 🤔  

Looks like it's unlikely that the `?` desugaring will change any time soon - https://github.com/rust-lang/rust/issues/31436#issuecomment-619427209

File: src/conversions/std/array.rs [hunk 12]
```diff
@@ -37,33 +36,38 @@ where
...
+    T: FromPyObjectOwned<'py>,
 {
     // Types that pass `PySequence_Check` usually implement enough of the sequence protocol
     // to support this function and if not, we will only fail extraction safely.
     let seq = unsafe {
         if ffi::PySequence_Check(obj.as_ptr()) != 0 {
             obj.cast_unchecked::<PySequence>()
         } else {
-            return Err(DowncastError::new(obj, "Sequence").into());
+            return Err(DowncastError::new_from_borrowed(obj, "Sequence").into());
```
Comment (davidhewitt) [inline src/conversions/std/array.rs:60]: We probably need to make `DowncastError::new_from_borrowed` public to make it usable for others. Or maybe we can make `DowncastError:new` generic on `Into<Borrowed<'_, '_, PyAny>>`.

I also have been wondering if we should call it `CastError` to match the function name.

Probably both are potential follow ups.

**Comment by davidhewitt** [comment]
Oh, and I think there's a note in the guide about performance of `.cast()` vs `.extract()` which is presumably not true once we've landed this change?

File: src/conversions/indexmap.rs (hunk 11 above)
Comment (Icxolu) [reply src/conversions/indexmap.rs:147]: Yep, it's the same problem as we had with `IntoPyObject`, really unfortunate... I would not be that bad if there would be a way to spell the `From` bound in the trait definition, but I don't think there is one, at least I could not fine one.

I will add a short note in the docs.

File: src/conversion.rs (hunk 10 above)
Comment (Icxolu) [reply src/conversion.rs:371]: Agreed, will add a short sentence here.

File: src/conversions/std/array.rs (hunk 12 above)
Comment (Icxolu) [reply src/conversions/std/array.rs:60]: #5394

File: guide/src/migration.md (hunk 6 above)
Comment (Icxolu) [reply guide/src/migration.md]: Yeah, this seems to creep up quite often 😅. I gave it a go, let me know what you think.

File: guide/src/migration.md (hunk 5 above)
Comment (Icxolu) [reply guide/src/migration.md]: Yeah, makes sense this evolved a bit over time 😁. I added a paragraph about the error type. I also notices that I documented here in the past that `extract_bound` would be deprecated, it seems like this got lost over time. I think this is fine, implementors can fix it up with the other changes, and there shouldn't really be any callers to it directly. Do you agree? Otherwise I can also add it back.

**Comment by Icxolu** [comment]
Thanks for the review ❤️ . I've opened some sub-issues in #5390 to track various followups from here.

File: src/conversion.rs (hunk 9 above)
Comment (Icxolu) [reply src/conversion.rs:281]: Good idea, I added a section with these. Do you have a suggestion for an example type with lifetimes?

File: guide/src/migration.md (hunk 6 above)
Comment (davidhewitt) [reply guide/src/migration.md]: I think that's good, and "Python lifetime" is probably a reasonable name for the concept given it comes up so often 👍

File: guide/src/migration.md (hunk 5 above)
Comment (davidhewitt) [reply guide/src/migration.md]: Agree with all, looks great now 👍 

I also agree that given the trait has now changed so much, adding back `extract_bound` isn't likely to help at all.

File: src/conversion.rs (hunk 9 above)
Comment (davidhewitt) [reply src/conversion.rs:281]: Looking great!

For the type with lifetimes, I suggest using either `&'a str` or `Bound<'py, T>` and saying "this is roughly how this is implemented in PyO3", I think just choosing one of the two is good enough.

**Review by davidhewitt — APPROVED** [review]
Subject to final documentation fixups, let's merge!

Thank you for all the iteration and R&D done here 🚀

File: src/conversion.rs (hunk 9 above)
Comment (Icxolu) [reply src/conversion.rs:281]: Done 👍

**Comment by Icxolu** [comment]
Great, all green 🎉 I'm going to send this now. We can handle any touchups that are left in followups.
